srcDLL/DllMain.cpp

```diff
@@ -64,7 +64,7 @@ int __fastcall ExtInit(TApp *thisPtr, int)
 		PostErrorMessage(__FILE__, __LINE__, stringStream.str());
 	}
 
-	InstallIpDropDown();
+	ipDropDown.Initialize();
```

Comment:
Sorry, I wasn't very clear here. We seem to have a reversed concept of dependencies between static and member method.

One of the possibilities I was thinking of is:
```
auto gameModule = IpDropDown::Patch(); // Call static method which creates the object
```
As opposed to creating the object first, and then patching:
```
IpDropDown ipDropdown;
ipDropDown.Initialize();
```

Another possibility, different from the above, is to use an RAII technique:
```
if (IsInternalModuleRequested("IpDropDown")) {
    RegisterModule(new IpDropDown());  // Constructing the object activates the patch
}
```
In this case, the patching code does go in the constructor, and we have no need for the `Load()` method. (Contrary to the documentation we recently wrote about not patching from the constructor, because objects in the current design are constructed before they are enabled / registered).

My idea of using tuples may be a bit premature, and perhaps not well thought out. You can define them to hold arbitrary types, though they are still constructed at runtime with all types instantiated. Perhaps what I want is closer to what I saw in the Google Test documentation. I don't actually know how they do it, but clearly they were able to do it. It was the section on [Type Parameterized Tests](https://github.com/google/googletest/blob/master/googletest/docs/advanced.md#type-parameterized-tests):

> Finally, you are free to instantiate the pattern with the types you want. If you put the above code in a header file, you can #include it in multiple C++ source files and instantiate it multiple times.
> ```
> typedef ::testing::Types<char, int, unsigned int> MyTypes;
> INSTANTIATE_TYPED_TEST_SUITE_P(My, FooTest, MyTypes);
> ```
> 

Perhaps it may require evil macros to accomplish.

Reply:
OK, the constructor portion makes sense now. I agree there is nothing stopping us from executing the Load function from the constructor and the unload in the destructor. This would probably involve removing the virtual Load and Unload function from GameModule and instead executing its code in the constructor and destructor. There is a certain symmetry to this solution.

I'm happy with how things are, but am okay with reprogramming to to use constructor/destructor instead of virtual load and unload function. @DanRStevens would you like to see it changed constructor/destructor? Shouldn't be too hard...

---

OK, I see what you mean that we could use a static function to create the object and return it, but I don't feel like it adds much here?

---

For the tuple information, I think what you are describing looks generally good to me, but it is still fuzzy. I sort of understand the idea of tuples based on some C# work I've done, but suspect there are differences in how C++ implements. I like the idea of punting for now on this for now in an issue or just waiting to see if it becomes useful later.

Reply:
Yes, this might be more of a longer term plan. It would require messing around with templates, which can sometimes be tricky to get right.

Moving from `Load`/`Unload` to constructor/destructor patching shouldn't be too difficult of a change, from the module perspective, so it's fine if we punt on that too.

Actually, to really make use of the constructor/destructor patching, we could design some RAII type patch objects. Such objects could patch/unpatch common instructions in their constructor/destructor. That would allow a very declarative syntax to patch calls and jumps, which are the majority of patch code. Example:
```
DisableCdCheckPatch : GameModule {
private:
  JumpNop<0xsomeaddr> neverJumpToCdCheckCode;
  JumpUnconditional<0xsomeaddr> alwaysSkipCdCheckCode;
};

GarageSaveLoadPatch : GameModule {
public:
  // Constructor
  GarageSaveLoadPatch() :
    // Initialize RAII patch objects
    preSaveTranslate(&GarageSaveLoadPatch::NewPreSaveTranslate);
  {}

private:
  bool NewPreSaveTranslate(params) {
    // Patch code ...
  };

  CallRelink<0xsomeaddr> preSaveTranslate;
};
```